### DISCUSSION TITLE
SetRemoteDescription called with no ice-ufrag

### DISCUSSION THREAD

**Discussion by chinatsm** [pion/webrtc#1817] [Q&A] [answered]
Hello, I just learned webrtc recently, this is great, I’m a little confused, please help me answer
The remote client Javascript generates the following SDP
  {
sdp: "v=0
↵o=- 3430854848718527786 2 IN IP4 127.0.0.1
↵s=-
↵t=0 0
↵a=msid-semantic: WMS
↵",
type: "offer",
}

Through SetRemoteDescription to handle the error as follows, may I ask what went wrong
SetRemoteDescription called with no ice-ufrag

**Comment by Sean-Der** [comment, marked as answer]
Hey @chinatsm 

You need to add at least one track on create one datachannel before generating an Offer. This is true for establing a WebRTC connection between any implementation.

thanks for using Pion!

**Reply by chinatsm** [reply]
Thank you very much for your reply
Now is to do a p2p network to watch the video, read the video stream from the Pion backend, and then watch it through the browser,
The browser transmits the signal through websocket, so a data channel and an offer should be created on the browser side first? So there is ice-ufrag, right?
Are there any examples of this type of Javascript, thank you

**Reply by Sean-Der** [reply]
You could either have the server generate the offer

* Connect to WebSocket
* First message from WebSocket would be offer from Pion server
* Send answer via WebSocket
* Exchange ICE Candidates as they trickle

Or you can generate two recvony transceivers and then generate the offer.

```
let pc = new RTCPeerConnection()

pc.addTransceiver('video', {direction: 'recvonly'})
pc.addTransceiver('audio', {direction: 'recvonly'})
```

**Reply by chinatsm** [reply]
Thank you for your help
Now I can receive ice-ufrag, but the Answer received in Pion cannot be used, and the error "invalid proposed signaling state transition: stable->SetRemote(answer)->stable" is reported.
The following is the code：
go
```go 
func (this *Peer) Offer(id string) (info ConnectInfo, err error) {

	offer, err := this.Pc.CreateOffer(nil)
	if err != nil {
		log.Println("offer create offer error:", err)
		this.Close()
		return
	}
	if err = this.Pc.SetLocalDescription(offer); err != nil {
		log.Println("offer rtc error:", err)
		this.Close()
		return
	}
	info = ConnectInfo{
		Description: Description{
			Type: "offer",
			Sdp:  offer.SDP,
		},
		From: "server",
		To:   id,
	}
	return
}
```

javascript
```javascript
if (description) {
                try {
                    await pc.setRemoteDescription(description);
                } catch(err) {
                    await Promise.all([
                        pc.setLocalDescription({type: "rollback"}),
                        pc.setRemoteDescription(description)
                    ]);
                } finally {
                    if (description.type == "offer") {
                        await pc.setLocalDescription(await pc.createAnswer());
                        signaler.send({description: pc.localDescription, from: to , to: from});
                    }
                }
            } else if (candidate) {
                await pc.addIceCandidate(candidate);
            }
```

go
 ```go
case "answer":
		anser := webrtc.SessionDescription{}
		if err = json.Unmarshal([]byte(tojson(msg.Description)), &anser); err != nil {
			return
		}
		err = this.Pc.SetRemoteDescription(anser)
		if err != nil {
			return
		}
```
<img width="853" alt="tupian" src="https://user-images.githubusercontent.com/49516465/119078932-c6ffa400-ba29-11eb-8a55-4d5fa3d321bc.png">

**Reply by Sean-Der** [reply]
That error happens with `CreateOffer` and `SetLocalDescription` hasn't been called. 

Is it possible that you have multiple PeerConnections and are setting the answer on one that hasn't been used yet? I am happy to help, but would you prefer to discuss this here or on [Stackoverflow](https://stackoverflow.com/questions/67617458/report-an-error-invalid-proposed-signaling-state-transition-stable-setremote) just don't want to duplicate the effort.

**Reply by chinatsm** [reply]
Okay, I'm very sorry, I have been deleted from StackOverflow, thank you very much for your help, there are now two PEERs
One is PION and the other is a browser.
PION is used to read the local stream, and the browser can watch it through websocket communication
Is this possible?

**Reply by chinatsm** [reply]
```
func (this *Peer) Offer() (offer webrtc.SessionDescription, err error) {

	offer, err = this.Pc.CreateOffer(nil)
	if err != nil {
		log.Println("offer create offer error:", err)
		this.Close()
		return
	}
	err = this.Pc.SetLocalDescription(offer)
	if err != nil {
		log.Println("offer rtc error:", err)
		this.Close()
		return
	}
	return
}
```
Hello, I carefully checked the code according to your suggestion, and still reported an error after using SetLocalDescription: InvalidModificationError: The suggested signaling state transition is invalid: stable-> SetRemote(answer)-> stable
The current situation is:
There is no problem sending the Offer from the browser -> Pion receives and sends the Answer and the created Offer to the browser. That is, at the end Pion received the browser's Answer signal and reported an error when setting the RemoteDescription.
I haven’t solved it after testing for a long time, please help me, thank you

**Reply by Sean-Der** [reply]
Hey @chinatsm 

Sorry I am not sure exatly. I would maybe start with [sfu-ws]( https://github.com/pion/example-webrtc-applications/tree/master/sfu-ws) and modify it.
